### DISCUSSION TITLE
Upcoming Version 4 Discussion

### DISCUSSION THREAD

**Discussion by JamesIves** [JamesIves/github-pages-deploy-action#524] [General]
# Version 4 Beta 🚨🚨🚨

Version  4 has now been released. Please refer to the README! 
 
---
---
---

This thread outlines some of the plans for version 4. You can point your workflow to `releases/v4` to try it early, and [you can find the dev branch here](https://github.com/JamesIves/github-pages-deploy-action/tree/dev-v4).

Latest release changes (Look for V4 prereleases): https://github.com/JamesIves/github-pages-deploy-action/releases

## Implemented Changes

* Obsoleted `LFS` and `PRESERVE`: https://github.com/JamesIves/github-pages-deploy-action/pull/515 
* ESLint for tests: https://github.com/JamesIves/github-pages-deploy-action/pull/517 
* dry-run capabilities: https://github.com/JamesIves/github-pages-deploy-action/pull/505 
* Remove the need for the `GITHUB_TOKEN` input variable: https://github.com/JamesIves/github-pages-deploy-action/issues/525
* CodeQL Adjustments: https://github.com/JamesIves/github-pages-deploy-action/pull/540
* PR based integration tests:  https://github.com/JamesIves/github-pages-deploy-action/pull/505
* Automate deployment task: https://github.com/JamesIves/github-pages-deploy-action/pull/541
* Automate release branch deployments.
* Improve the CLEAN_EXCLUDE user experience
* Make input variables lowercase in the documentation to be more consistent with other actions. (Needs to be adjusted in `action.yml` before release is cut)
* Hyphenate arguments instead of lower_snakes
* Point integration tests to the `release/v4` branch in `dev-v4`.
* Update security support docs.
* Support SSH key configuration natively: https://github.com/JamesIves/github-pages-deploy-action/issues/459 (https://github.com/JamesIves/github-pages-deploy-action/pull/569)
* Add dev support for GitHub Codespaces.
* Prevent previously set git configs from interfering with deployment. 
* Overall README sweep.
* Improved logo/artwork.
* Fully deprecate and drop v1 support via the `master` branch.

## Discussion 
If you have any suggestions/ideas please make a separate thread for consideration. The plan is to release version 4 at the end of the year.

**Comment by Pike** [comment]
About the README:

The github actions use lower-case arguments in documentation, should we do so here, too? `getInput` is case-insensitive, so this should really just be a documentation issue?

PS: Thanks for the merges ;-)

**Reply by JamesIves** [reply]
I like it, let's make these lowercase.

Related issue: https://github.com/JamesIves/github-pages-deploy-action/issues/531

**Comment by Pike** [comment]
About integration tests:

With integration tests running (and hopefully passing with dry-run), can `integration-beta.yml` and its corresponding branch go away?

**Reply by JamesIves** [reply]
I think so yes! That way we have coverage on pull requests, and daily scheduled checks on the live release branch.

**Comment by Pike** [comment]
One more thing that I'd suggest for v4:

Deploy production node modules on the release branch only.

The existing release branch already has dev dependencies, too, but that'd be easy to fix.

My rationale is really just personal taste, I like to keep production environments tidy.

**Reply by JamesIves** [reply]
Agreed. I'd actually like to automate these deployments via a workflow job. GitHub Actions lets you manually trigger workflows now, which would remove the need to manually switch branches and actually run the yarn commands manually. So yeah, if this was automated it would resolve a lot of minor issues like this.

It would need to do the following
- Push dev branch to `release/v4` (We can just use the deploy action to commit this)
- Checkout release branch, run the yarn install command and commit the prod dependencies.

And then once the job is done, all that would be left is to tag the release branch and write the changelog, and it would then handle the rest via the publish task. Might as well go full automation as that's the only manual part left in the process lol.

**Reply by Pike** [reply]
What do you  think about https://github.com/JamesIves/github-pages-deploy-action/compare/releases/v4...Pike:releases/v4 ?

**Reply by JamesIves** [reply]
Looks good to me! 🎉

**Reply by Pike** [reply]
Created #541 for that change

**Comment by JamesIves** [comment]
Regarding CLEAN_EXCLUDE UX: 

I think it might be a better idea to have a comma seperated list, ie `.nojekyll, .gitconfig` as opposed to a string formatted as an array. It feels more natural to do this as we're sort of breaking the norms of an array by expecting every item to be a string, plus it just feels weird to type out like that. 

I'll likely make this change over the weekend if there's no opposition.

**Reply by Pike** [reply]
https://github.com/actions/toolkit/issues/184 mentions that upstream github actions use multiline strings as a workaround.

Maybe that's more future-proof? Also, avoids the problem of files with `,` 😄

**Reply by Pike** [reply]
I've thrown up a PR for this, #553 .

**Reply by JamesIves** [reply]
Sorry - will get to this soon. Still not feeling great. Probably after this weekend I'll be able to.

**Comment by Pike** [comment]
Should `dev-v4` be the default branch now? In particular now that changes to the scheduled integration tests landed on it, those only start taking effect when it is.

**Reply by JamesIves** [reply]
Not just yet, I want to get through the major features in version 4 before we switch over the branches. I'd like to keep version 3 as the currently supported branch until then. 

Probably this weekend I'll switch them over. 👍🏻

**Reply by JamesIves** [reply]
Hoping to wrap up the SSH changes sometime early next week, spent a good amount of time on it this weekend and I'm just testing a few last things before I will have a PR ready. Once that part is ready we can switch things over as that's the last major change in the release.

**Reply by JamesIves** [reply]
PR: https://github.com/JamesIves/github-pages-deploy-action/pull/569

**Reply by JamesIves** [reply]
releases/v4 has been updated and a new beta version has been cut - I've put a note at the top of the thread calling for people to test this. Gonna push this up next weekend after testing and the remaining housekeeping tasks are done!

**Reply by JamesIves** [reply]
Aiming to push out v4 this week! 🎉

**Comment by Pike** [comment]
Is #567 about external repos a blocker? It seems that I broke pushing to external repos with the worktree patch?

**Reply by JamesIves** [reply]
Yeah, I'm going to convert it to an issue to make sure it's tracked. I'll also do some testing to ensure that this is an actual issue: https://github.com/JamesIves/github-pages-deploy-action/issues/578

I also did a bunch of testing this evening ahead of the release and I've found a few blockers with repro steps which I've listed out here: https://github.com/JamesIves/github-pages-deploy-action/issues?q=is%3Aissue+is%3Aopen+label%3Ablocker
